### PR TITLE
Fix clang compiler warnings

### PR DESCRIPTION
**Pull request by olehermanse**

### CODE REVIEW

**Comment by olehermanse** [comment]
@jimis Please review. This is the first round of warnings, from just building with clang, no scan-builder.

File: cf-serverd/server_classic.c [hunk 1]
```diff
@@ -1083,7 +1083,7 @@ int BusyWithClassicConnection(EvalContext *ctx, ServerConnectionState *conn)
     case PROTOCOL_COMMAND_AUTH_PLAIN:
         SetConnectionData(conn, (char *) (recvbuffer + strlen("CAUTH ")));
 
-        if (conn->username == NULL || !IsUserNameValid(conn->username))
```
Comment (jimis) [inline cf-serverd/server_classic.c]: This is not bad code, just overcautious code. Now the compiler will optimise this away. But if in the future `username` becomes a pointer from an array, this check will be needed.

**Comment by jimis** [comment]
Good job, can you introduce a clang parallel build with clang?

**Comment by olehermanse** [comment]
Added 3 more fixes as separate commits. These were reported by clang scan-builder.

A new clang job has been added to Travis. I also checked that it fails if I introduce a warning on purpose: https://travis-ci.org/cfengine/core/builds/340506930

Jenkins build: https://ci.cfengine.com/view/all/job/pr-pipeline/76/

@jimis Please review and merge when green.

**Comment by olehermanse** [comment]
@jimis It's green.

File: libutils/unix_dir.c [hunk 2]
```diff
@@ -194,8 +194,10 @@ static size_t GetDirentBufferSize(size_t name_len)
 
 struct dirent *AllocateDirentForFilename(const char *filename)
 {
-    struct dirent *entry = xcalloc(1, GetDirentBufferSize(strlen(filename)));
+    int length = strlen(filename);
+    struct dirent *entry = xcalloc(1, GetDirentBufferSize(length));
 
-    strcpy(entry->d_name, filename);
+    // d_name is fixed length, but we have allocated extra space using xcalloc:
+    strncpy((char *)entry->d_name, filename, length + 1);
```
Comment (jimis) [inline libutils/unix_dir.c]: `strncpy` is generally considered bad practice for string copying, slow and easy to get it wrong. Here all you need is `memcpy` or `strcpy`, or if you want to be extra pedantic go with `strlcpy`.

Comment (jimis) [reply libutils/unix_dir.c]: The "fix" here is basically the cast, so add to your comment that this cast is used to silence compiler warnings.

File: travis-scripts/script.sh [hunk 3]
```diff
@@ -39,7 +39,7 @@ then
     make CFLAGS=-Werror
 elif [ "$JOB_TYPE" = compile_and_unit_test ]
 then
-    make CFLAGS=-Werror  &&
+    make CFLAGS="-Werror -Wno-pointer-sign -Wno-duplicate-decl-specifier"  &&
```
Comment (jimis) [inline travis-scripts/script.sh]: Can you add an explanation to the commit message why it's OK to skip these two? Or maybe some TODO comment in the travis file to fix that in the future?

File: libutils/unix_dir.c (hunk 2 above)
Comment (olehermanse) [reply libutils/unix_dir.c]: Thanks, used memcpy instead.

File: travis-scripts/script.sh (hunk 3 above)
Comment (olehermanse) [reply travis-scripts/script.sh]: Thanks, added.

**Comment by olehermanse** [comment]
https://ci.cfengine.com/view/all/job/pr-pipeline/80/

**Comment by olehermanse** [comment]
Builds look green, merging.

**Comment by jimis** [comment]
+1

**Pull request merged**